api/app/auto_spatial_advisory/polygonize.py

```diff
@@ -0,0 +1,52 @@
+""" Code for polygonizing a geotiff file. """
+from typing import Tuple
+from osgeo import gdal, ogr, osr
+import numpy as np
+
+
+def _create_in_memory_band(data: np.ndarray, cols, rows, projection, geotransform):
+    """ Create an in memory data band to represent a single raster layer.
+    See https://gdal.org/user/raster_data_model.html#raster-band for a complete
+    description of what a raster band is.
+    """
+    mem_driver = gdal.GetDriverByName('MEM')
+
+    dataset = mem_driver.Create('memory', cols, rows, 1, gdal.GDT_Byte)
+    dataset.SetProjection(projection)
+    dataset.SetGeoTransform(geotransform)
+    band = dataset.GetRasterBand(1)
+    band.WriteArray(data)
+
+    return dataset, band
+
+
+def polygonize(geotiff_filename) -> Tuple[ogr.DataSource, ogr.Layer]:
+    """  Given some tiff file, return a polygonized version of it, in memory, as an ogr layer. """
+    source: gdal.Dataset = gdal.Open(geotiff_filename, gdal.GA_ReadOnly)
+
+    source_band = source.GetRasterBand(1)
+    source_data = source.ReadAsArray()
+    # https://gdal.org/api/python/osgeo.osr.html#osgeo.osr.SpatialReference
+    spatial_reference: osr.SpatialReference = source.GetSpatialRef()
+
+    # generate mask data
+    mask_data = np.where(source_data == 0, False, True)
+    mask_ds, mask_band = _create_in_memory_band(
+        mask_data, source_band.XSize, source_band.YSize, source.GetProjection(),
+        source.GetGeoTransform())
+
+    # Create a memory OGR datasource to put results in.
+    mem_drv: ogr.Driver = ogr.GetDriverByName("Memory")
```

Comment:
Is there a diff between using the "Memory" driver and the "MEM" driver in line 12?

Reply:
You're making me go DEEP into the horrid documentation. 

`gdal.GetDriverByName('MEM')` is asking for a in memory raster driver: https://gdal.org/drivers/raster/mem.html

If you do:
`
mem_driver = gdal.GetDriverByName('MEM')
dataset = mem_driver.Create('memory', cols, rows, 1, gdal.GDT_Byte)
`
then `memory` is just a name - no meaning.

If you do
`ogr.GetDriverByName("Memory")` 
then it's asking for an in memory vector driver: https://gdal.org/drivers/vector/memory.html#vector-memory

Reply:
I've added comments.